### PR TITLE
Feature/authorization registration

### PR DESCRIPTION
**Pull request by tresdosdos**

### CODE REVIEW

File: client/src/di-container/services/http/http.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      .then(res => res.json())
+      .then(this.handleError);
+  }
+
+  public delete(url, options?) {
+    return fetch(
+      url,
+      Object.assign(
+        {
+          headers: this.headers,
```
Comment (lightness) [inline client/src/di-container/services/http/http.tsx]: Can `delete` method have body?

File: client/src/di-container/services/user/user-token.interface.tsx [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { User } from './user.interface';
+
+export interface UserAndToken {
+  data: User;
```
Comment (lightness) [inline client/src/di-container/services/user/user-token.interface.tsx]: `data`? Try to assign more specific field titles

File: client/src/di-container/stores/spinner.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import { observable } from 'mobx';
+
+import { SpinnerStoreModel } from '../interfaces';
+
+@injectable()
+export class SpinnerStore implements SpinnerStoreModel {
+  @observable
+  public spinner: boolean = false;
+
+  public start() {
```
Comment (lightness) [inline client/src/di-container/stores/spinner.tsx]: You need to mark it with `@action` decorator

File: client/src/di-container/stores/user.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  constructor(@inject(TYPES.UserService) private userService: UserServiceModel) {
+    this.userService
+      .get()
+      .then((user: User) => {
+        this.userData = user;
+      })
+      .catch(() => {
+        storageService.resetToken();
+      });
```
Comment (lightness) [inline client/src/di-container/stores/user.tsx]: async/await

File: client/src/di-container/stores/user.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  public get user(): User {
+    return this.userData;
+  }
+
+  public set user(info: User) {
```
Comment (lightness) [inline client/src/di-container/stores/user.tsx]: transform to method and mark with `@action` decorator

File: client/src/di-container/types.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export const TYPES = {
+  HttpService: Symbol.for('HttpServiceModel'),
```
Comment (lightness) [inline client/src/di-container/types.tsx]: Why use `Symbol.for()`? 
`Symbol()` is enough...

File: client/src/env.tsx [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export const environment = {
```
Comment (lightness) [inline client/src/env.tsx]: why file is `.tsx`? And not `.ts`?

File: server/src/config/config.default.ts [hunk 8]
```diff
@@ -1,5 +1,23 @@
...
-  port: 2000
+  port: 2000,
+  db: {
+    user: process.env.DB_USER,
+    password: process.env.DB_PASSWORD,
+    host: process.env.DB_HOST,
+    database: process.env.DB_NAME,
+    port: process.env.DB_PORT
+  },
+  jwtSecrets: {
```
Comment (lightness) [inline server/src/config/config.default.ts]: Why use different secrets?

File: server/src/container/interfaces/email-sdk.interface.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import { MailOptions } from '../tools/email';
+
+export interface EmailSdkModel {
```
Comment (lightness) [inline server/src/container/interfaces/email-sdk.interface.ts]: It isn't model...
Try to name interfaces differently... May be prefix `I` is not so bad :)

File: server/src/container/inversify.config.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { Container } from 'inversify';
+import { TYPES } from './types';
+import { JwtSdkModel, UserServiceModel, JwtServiceModel, EmailSdkModel, EmailServiceModel } from './interfaces';
+import { EmailService, JWTService, UserService } from './services';
+import { JwtSdk } from './tools/jwt';
+import { EmailSdk } from './tools/email';
+
+const container = new Container();
+container.bind<UserServiceModel>(TYPES.UserService).to(UserService);
```
Comment (lightness) [inline server/src/container/inversify.config.ts]: As I remember you can do just 
```ts

container.bind<UserService>(TYPES.UserService).to(UserService);
```

File: server/src/container/inversify.config.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+container.bind<JwtSdkModel>(TYPES.JwtSdk).to(JwtSdk);
+container.bind<JwtServiceModel>(TYPES.JwtService).to(JWTService);
+container.bind<EmailSdkModel>(TYPES.EmailSdk).to(EmailSdk);
+container.bind<EmailServiceModel>(TYPES.EmailService).to(EmailService);
+
+const userService = container.get<UserServiceModel>(TYPES.UserService);
+const jwtService = container.get<JwtServiceModel>(TYPES.JwtService);
+const emailService = container.get<EmailServiceModel>(TYPES.EmailService);
```
Comment (lightness) [inline server/src/container/inversify.config.ts]: You don't need do that

Comment (lightness) [reply server/src/container/inversify.config.ts]: Using `container.get` isn't good practice. Try to get rid

File: server/src/container/services/email/index.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import { inject, injectable } from 'inversify';
+
+import config from '../../../config';
+import * as _ from 'lodash';
```
Comment (lightness) [inline server/src/container/services/email/index.ts]: Optimize imports

File: server/src/container/services/email/templates/registration-text.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+export const REGISTRATION_TEXT = `Welcome to Poll-app!
+    Dear user! Thanks you for registering on Poll-app.
+    To activate your account you should go to registration link.
+    Your registration link - <%= link %>
+    <%= additionalText %>`;
```
Comment (lightness) [inline server/src/container/services/email/templates/registration-text.ts]: you can use some template engine... Storing long texts in ts doesn't seems ok)

File: server/src/container/services/jwt.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { inject, injectable } from 'inversify';
+
+import config from '../../config';
+
+import { TYPES } from '../types';
+import { JwtSdkModel, JwtServiceModel } from '../interfaces';
+
+@injectable()
+export class JWTService implements JwtServiceModel {
```
Comment (lightness) [inline server/src/container/services/jwt.ts]: `JwtService`

File: server/src/container/services/user.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
+import * as bcrypt from 'bcrypt';
+
+import { injectable } from 'inversify';
+
+import User from '../../db/models/user.model';
+
+import { UserModel } from '../../models/user';
```
Comment (lightness) [inline server/src/container/services/user.ts]: Something went wrong with your imports. Please, check all files

File: server/src/container/services/user.ts [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+  public async getAndCompare(user) {
+    const userData = await User.findOne({
+      where: {
+        email: user.email
+      }
+    });
+    if (!userData) {
+      throw new InvalidCredentialsError();
+    }
```
Comment (lightness) [inline server/src/container/services/user.ts]: Try to leave empty line before and after `if`... Code will be much more readable

File: server/src/container/services/user.ts [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+import User from '../../db/models/user.model';
+
+import { UserModel } from '../../models/user';
+import { InvalidCredentialsError, UserExistsError, InvalidEmailError } from '../../utils/errors';
+import { UserServiceModel } from '../interfaces';
+
+@injectable()
+export class UserService implements UserServiceModel {
+  public async getById(id) {
+    const user = await User.findOne({ where: { id: id } });
```
Comment (lightness) [inline server/src/container/services/user.ts]: `where: { id }`

File: server/src/container/services/user.ts [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    return this.build(user.dataValues);
+  }
+
+  public async getByEmail(email) {
+    const user = await User.findOne({ where: { email } });
+    return this.build(user.dataValues);
+  }
+
+  public async getAndCompare(user) {
+    const userData = await User.findOne({
```
Comment (lightness) [inline server/src/container/services/user.ts]: try to name variables more meaningful

File: server/src/container/services/user.ts [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+  private build(user): UserModel {
+    return {
+      id: user.id,
+      email: user.email,
+      confirmed: user.confirmed,
+      createdAt: user.createdAt,
+      updatedAt: user.updatedAt
+    };
+  }
```
Comment (lightness) [inline server/src/container/services/user.ts]: Why we need this method? Method title doesn't correlate with content

File: server/src/routes/user.ts [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    } catch (err) {
+      next(err);
+    }
+  } else {
+    res.send(null);
+  }
+});
+
+router.get('/:id', async (req: Request, res: Response) => {
+  userService.getById(req.params.id).then((user: UserModel) => res.send(user));
```
Comment (lightness) [inline server/src/routes/user.ts]: async method... But why `then`?

File: server/src/routes/user.ts [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+router.put('/reset', async (req: Request, res: Response, next: Function) => {
+  const { email } = req.body;
+  try {
+    const user = await userService.getByEmail(email);
+    if (user) {
+      await emailService.sendResetMail({ email });
+      const msg = `Reset email was delivered to ${email}`;
+      logger.info(msg);
+      res.send({ message: msg });
+    }
```
Comment (lightness) [inline server/src/routes/user.ts]: Try following:
```ts
if (!user) {
  return;
}

await emailService.sendResetMail({ email });
const msg = `Reset email was delivered to ${email}`;
logger.info(msg);
res.send({ message: msg });
```

Comment (lightness) [reply server/src/routes/user.ts]: code reads easier, agree?

File: client/src/di-container/stores/user.tsx (hunk 4 above)
Comment (tresdosdos) [reply client/src/di-container/stores/user.tsx]: sorry, but it is a constructor

File: server/src/services/user.ts [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  public async activate(user) {
+    const { id } = user;
+    const updateNumber = await User.update({ confirmed: true }, { where: { id } });
+    if (updateNumber[0] === 0) {
+      throw new JsonWebTokenError();
+    } else {
+      const foundUser = await User.scope('public').findOne({ where: { id } });
+      return foundUser.dataValues;
```
Comment (lightness) [inline server/src/services/user.ts]: As I remember, sequelize have `.toJSON` method for this purpose

File: server/src/services/user.ts [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+          throw new InvalidEmailError();
+        }
+        case 'SequelizeUniqueConstraintError': {
+          throw new UserExistsError();
+        }
+      }
+    }
+  }
+
+  public async activate(user) {
```
Comment (lightness) [inline server/src/services/user.ts]: Do following:
1. Find by id
2. Update on instance (not on model)
3. Just return

File: server/src/services/user.ts [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+      throw new JsonWebTokenError();
+    } else {
+      const foundUser = await User.scope('public').findOne({ where: { email } });
+      return foundUser.dataValues;
+    }
+  }
+
+  private hashPassword(password: string): string {
+    const SALT_WORK_FACTOR = Math.random();
+    return bcrypt.hash(password, SALT_WORK_FACTOR);
```
Comment (lightness) [inline server/src/services/user.ts]: As I remember `bcrypt` can generate salt by itself

File: server/src/tools/email/index.ts [hunk 25]
```diff
@@ -0,0 +1,30 @@
+import * as emailjs from 'emailjs';
+import * as bluebird from 'bluebird';
+
+import { injectable } from 'inversify';
+
+import config from '../../config';
+
+import { IEmailSdk } from '../../container';
```
Comment (lightness) [inline server/src/tools/email/index.ts]: Something wrong!!!

File: server/src/tools/jwt/index.ts [hunk 26]
```diff
@@ -0,0 +1,29 @@
+import * as jwt from 'jsonwebtoken';
+import { injectable } from 'inversify';
+
+import { IJwtSdk } from '../../container';
+import { TokenExpiredError, JsonWebTokenError } from '../../utils/errors';
+
+@injectable()
+export class JwtSdk implements IJwtSdk {
+  public generateToken(payload: object, expIn, tokenSecret, alg = 'HS256') {
```
Comment (lightness) [inline server/src/tools/jwt/index.ts]: No need to get `tokenSecret` from outside.
Just import it here

File: server/src/container/interfaces/user-service.interface.ts [hunk 27]
```diff
@@ -0,0 +1,12 @@
+import { UserModel } from '../../models';
+import { FilteredModelAttributes } from 'sequelize-typescript/lib/models/Model';
```
[the remark this answers is not in the document]
Comment (tresdosdos) [reply server/src/container/interfaces/user-service.interface.ts]: no, sorry

File: client/src/components/login/index.tsx [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+  email: string;
+  password: string;
+  emailError?: string;
+  passwordError?: string;
+}
+
+export class Login extends React.Component<LoginProps, LoginState> {
+  private formValidator = FormValidator.instance;
+
+  constructor(props: any) {
```
Comment (lightness) [inline client/src/components/login/index.tsx]: any?

File: client/src/components/login/index.tsx [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+    super(props);
+    this.state = {
+      email: '',
+      password: ''
+    };
+  }
+
+  public handleEmailChange = (e: any) => {
```
Comment (lightness) [inline client/src/components/login/index.tsx]: react-typescript provide typings for it

File: client/src/containers/home.tsx [hunk 30]
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+
+export class Home extends React.Component {
+  public render() {
+    return <button>create</button>;
+  }
+}
```
Comment (lightness) [inline client/src/containers/home.tsx]: Your home seems strange

File: client/src/services/http/http.ts [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+          method
+        },
+        options
+      )
+    );
+    const parseResult = await fetchResult.json();
+    return this.handleError(parseResult);
+  }
+
+  private handleError(res: any): Promise<any> {
```
Comment (lightness) [inline client/src/services/http/http.ts]: any?

File: client/src/services/http/http.ts [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+    if (res.statusCode) {
+      throw res;
```
Comment (lightness) [inline client/src/services/http/http.ts]: you should not throw with object!

File: client/src/services/message.ts [hunk 33]
```diff
@@ -0,0 +1,23 @@
+import { message } from 'antd';
+
+export class MessageService {
+  private static serviceInstance = new MessageService();
+
+  private constructor() {}
+
+  public static get instance(): MessageService {
+    return this.serviceInstance;
```
Comment (lightness) [inline client/src/services/message.ts]: Don't use `this` in static members!!! It's only typescript feature... And it's very doubtful!

File: client/src/services/token.ts [hunk 34]
```diff
@@ -0,0 +1,22 @@
...
+
+  private constructor() {}
+
+  public static get instance() {
+    return this.serviceInstance;
+  }
+
+  public getToken(): string {
+    const token = localStorage.getItem('access_token');
+    return token ? token : '';
```
Comment (lightness) [inline client/src/services/token.ts]: 1. empty line before return.
2. `token || ''`

File: client/src/services/user.ts [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+
+  public create(user: UserModel): Promise<UserAndToken> {
+    return this.httpService.post(`${environment.baseUrl}/user`, user);
+  }
+
+  public get(userId?: number): Promise<UserModel> {
+    if (!userId) {
+      return this.httpService.get(`${environment.baseUrl}/user/`);
+    }
+    return this.httpService.get(`${environment.baseUrl}/user/${userId}`);
```
Comment (lightness) [inline client/src/services/user.ts]: empty line

File: client/src/services/user.ts [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+    return this.serviceInstance;
+  }
+
+  public create(user: UserModel): Promise<UserAndToken> {
+    return this.httpService.post(`${environment.baseUrl}/user`, user);
+  }
+
+  public get(userId?: number): Promise<UserModel> {
+    if (!userId) {
+      return this.httpService.get(`${environment.baseUrl}/user/`);
```
Comment (lightness) [inline client/src/services/user.ts]: Don't use trailing slash

File: server/src/container/interfaces/user-service.interface.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (lightness) [reply server/src/container/interfaces/user-service.interface.ts]: okay

File: server/src/tools/jwt/payload.interface.ts [hunk 37]
```diff
@@ -0,0 +1,14 @@
+export interface JWTPayload {
```
Comment (lightness) [inline server/src/tools/jwt/payload.interface.ts]: better place all our data in single field `payload` for example

Don't mix your data fields with jwt inner

File: server/tsconfig.json [hunk 38]
```diff
@@ -2,6 +2,9 @@
   "compilerOptions": {
     "target": "es6",
     "module": "commonjs",
+    "moduleResolution": "node",
+    "lib": ["es6", "dom"],
```
Comment (lightness) [inline server/tsconfig.json]: dom? on server side?

**Pull request merged**